staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go

```diff
@@ -0,0 +1,273 @@
+func (d *DynamicKMSEncryptionConfigContent) runWorker() {
+	for d.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem processes file content when there is a message in the queue.
+func (d *DynamicKMSEncryptionConfigContent) processNextWorkItem() bool {
+	// key here is dummy item in the queue to trigger file content processing.
+	key, quit := d.queue.Get()
+	if quit {
+		return false
+	}
+	defer d.queue.Done(key)
+
+	err := d.loadEncryptionConfig()
+	if err == nil {
+		d.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	// add dummy item back to the queue to trigger file content processing.
+	d.queue.AddRateLimited(key)
+
+	return true
+}
+
+// loadEncryptionConfig processes the next set of content from the file.
+func (d *DynamicKMSEncryptionConfigContent) loadEncryptionConfig() error {
+	encryptionConfig, err := d.readEncryptionFileContent(d.filePath)
+	if err != nil {
+		return err
+	}
+	if len(encryptionConfig) == 0 {
+		return fmt.Errorf("encryption config file is empty %q", d.filePath)
+	}
+
+	// compute hash for the encryption config file content.
+	encryptionConfigHash := d.computeHash(encryptionConfig)
+
+	// check if encryptionConfig is different from the current. Do nothing if they are the same.
+	if !d.hasEncryptionConfigChanged(encryptionConfigHash) {
+		return nil
+	}
+
+	err = d.processEncryptionConfig()
+	if err != nil {
+		return err
+	}
+
+	// update local copy of recent config content once processing is successful.
+	d.encryptionConfig.Store(encryptionConfigHash)
+	klog.V(2).InfoS("Loaded new kms encryption config content", "name", d.name)
+
+	return nil
+}
+
+func (d *DynamicKMSEncryptionConfigContent) readEncryptionFileContent(filePath string) ([]byte, error) {
+	return os.ReadFile(d.filePath)
+}
+
+// processEncryptionConfig processes the encryption config.
+func (d *DynamicKMSEncryptionConfigContent) processEncryptionConfig() error {
+	// get context to close the old transformers.
+	ctx, closeTransformers := wait.ContextForChannel(d.stopCh)
+
+	// ignoring content hash as it's already checked in loadEncryptionConfig.
+	transformerOverrides, healthChecks, _, maxTimeout, err := encryptionconfig.LoadEncryptionConfig(d.filePath, ctx.Done())
+	if err != nil {
+		klog.Warning("KMS encryption configuration is not valid.", err)
+		// close transformers to avoid leaking resources.
+		closeTransformers()
+		return err
+	}
+
+	// test if new transformers has good health.
+	err = d.validateNewTransformersHealth(healthChecks)
+	if err != nil {
+		klog.Warning("health check for new transformers failed.", err)
+		// close transformers to avoid leaking resources.
+		closeTransformers()
+		return err
+	}
+
+	// update transformers.
+	d.dynamicTransformers.Set(transformerOverrides, closeTransformers, maxTimeout)
+
+	return nil
+}
+
+func (d *DynamicKMSEncryptionConfigContent) validateNewTransformersHealth(healthChecks []healthz.HealthChecker) error {
```

Comment:
This needs to enforce a timeout, say `5 * time.Minute`.  Ideally this would be canceled early if the file hash changes but it should give at least `2 * time.Minute` to allow the check to complete in case the file is being mutated "often."